### PR TITLE
osd/scrub: Change scrub cost to average object size

### PR DESCRIPTION
**Pull request by amathuria**
osd_scrub_cost is currently set to 50_M which does not accurately represent the cost of scrub.
With mClock scheduler, we need the cost of an operation to represent the size of an operation. This change modifies scrub cost to the average of object sizes in a PG.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/osd/PG.cc
Comment (athanatos) [inline src/osd/PG.cc:467]: This duplicates logic in queue_recovery() -- factor it out into a helper (object_stat_sum_t::get_average_object_size() const)

**Review by athanatos — CHANGES_REQUESTED** [review]
AFAICT, this would cause queue_for_scrub, queue_scrub_after_repair, queue_for_rep_scrub, and queue_for_rep_scrub_reshed to be queued with a cost equal to the average object size.  That seems off as it should need to be multiplied by the number of objects to be scrubbed in that region.  Moreover, for shallow scrubs, we don't actually read the object contents so the cost would be independent of the object size.

Are we observing some behavior that requires a fix like this urgently?

**Comment by amathuria** [comment]
> 

@athanatos thanks for taking a look! That does make sense, I can discuss with you how we can adjust the cost for shallow and deep scrubs before making more changes. 
From the initial tests it looks like going ahead with the cost modelling change while keeping osd_scrub_cost as 50_M would considerably slow down deep scrubs. I am yet to verify this with shallow scrubs but I guess we would see a similar trend as the cost used is the same. 
I'll add the graphs from my testing here to clarify.

**Comment by amathuria** [comment]
> 

@athanatos thanks for taking a look! That does make sense, I can discuss with you how we can adjust the cost for shallow and deep scrubs before making more changes.
From the initial tests it looks like going ahead with the cost modelling change while keeping osd_scrub_cost as 50_M would considerably slow down deep scrubs. I am yet to verify this with shallow scrubs but I guess we would see a similar trend as the cost used is the same.
I'll add the graphs from my testing here to clarify.

File: src/osd/scrubber/pg_scrubber.h
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.h:809]: m_num_objects_current_chunk

File: src/osd/scrubber/pg_scrubber.cc
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.cc:1119]: This formatting change is irrelevant to this commit.  Either remove it or separate it into its own commit.

File: src/osd/scrubber/pg_scrubber.cc
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.cc:837]: Broken spacing.

File: src/osd/scrubber/pg_scrubber.cc
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.cc:847]: Spacing.  Please make a point of carefully reading over your PR in github prior to asking for a review.

File: src/osd/scrubber/pg_scrubber.cc
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.cc:848]: } else {

File: src/osd/scrubber/pg_scrubber.cc
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.cc:850]: shallow-scrub cost =

File: src/osd/scrubber/pg_scrubber.cc
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.cc:768]: This ceph_assert line now has trailing whitespace -- fix.

File: src/osd/scheduler/OpScheduler.h
Comment (athanatos) [inline src/osd/scheduler/OpScheduler.h:58]: Does this actually get executed?  This should probably have a ceph_assert(0 == "impossible for wpq"); check to ensure that it isn't actually invoked.

File: src/osd/PG.cc
Comment (athanatos) [inline src/osd/PG.cc:461]: Stray whitespace change.

File: src/osd/OSD.h
Comment (athanatos) [inline src/osd/OSD.h:606]: Most of the changes above are due to there only being two users that actually need to pass cost_per_object.  For the rest, add a queue_simple_scrub_event_msg variant that calls queue_scrub_event_msg with osd_scrub_event_cost.

File: src/osd/scrubber/pg_scrubber.cc
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.cc:839]: It seems like this is going to give a nonsense answer for wpq.  Seems like it should return osd_scrub_cost for wpq as it did before?

**Review by athanatos — CHANGES_REQUESTED** [review]
Approach seems reasonable.

File: src/osd/OSD.h
Comment (amathuria) [reply src/osd/OSD.h:606]: Yeah makes sense, I'll make the change

**Comment by amathuria** [comment]
jenkins test make check

**Comment by amathuria** [comment]
jenkins test make check

File: src/osd/OSD.cc
Comment (athanatos) [inline src/osd/OSD.cc:1746]: uint64_t

Comment (athanatos) [reply src/osd/OSD.cc:1746]: Audit other cases as well.

**Review by athanatos — CHANGES_REQUESTED** [review]
Needs rebase.

File: src/osd/PG.cc
Comment (sseshasa) [inline src/osd/PG.cc:422]: @amathuria Please note that this section is factored out to a method (see PG.h) as part of the snap trim cost PR (see https://github.com/ceph/ceph/pull/54597). This PR should get merged shortly. Therefore, you would need to rebase and call the new method here.

Comment (amathuria) [reply src/osd/PG.cc:422]: @sseshasa got it. I'll wait for that PR to get merged first

**Comment by amathuria** [comment]
jenkins test api

File: src/osd/PG.cc
Comment (amathuria) [reply src/osd/PG.cc]: @sseshasa does this look okay to you now?

Comment (sseshasa) [reply src/osd/PG.cc]: @amathuria Looks good to me.

**Comment by amathuria** [comment]
jenkins test api

**Comment by amathuria** [comment]
jenkins test windows

**Comment by amathuria** [comment]
jenkins test api

**Comment by amathuria** [comment]
@athanatos @sseshasa could you review this?

File: src/osd/scrubber/pg_scrubber.h
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.h:724]: This member is used to communicate a number of objects from PgScrubber::select_range() to PgScrubber::select_Range_n_notify().  Instead, modify select_range() to return the number of objects in the chunk (with 0 indicating that it's blocked).  Don't introduce a member  when a return value would suffice.

File: src/osd/OSD.h
Comment (athanatos) [inline src/osd/OSD.h:1620]: auto &&sdata isn't necessary -- use auto &sdata.  Remove the shard_index variable entirely and replace it with 0.

File: src/osd/scrubber/pg_scrubber.cc
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.cc:1559]: OSDService::queue_for_rep_scrub names this last parameter "cost_per_object".  Here, however, cost appears to be the total cost of the chunk already multiplied by the number of objects.  Please ensure that the names are consistent.

File: src/osd/OSD.cc
Comment (athanatos) [inline src/osd/OSD.cc:1764]: Do you actually mean cost_per_object here?  This appears to be the full cost of the event, not the per-object cost.  Presumably you mean event_cost?

Comment (amathuria) [reply src/osd/OSD.cc:1764]: Yes that's correct. I'll make the change to cost.

**Comment by amathuria** [comment]
jenkins test make check

**Comment by amathuria** [comment]
jenkins test api

**Comment by amathuria** [comment]
jenkins test api

**Comment by amathuria** [comment]
jenkins test make check

File: src/osd/OSD.cc
Comment (athanatos) [inline src/osd/OSD.cc:1817]: Why isn't this a queue_simple_scrub_event_msg?

Also, please move the rename from queue_scrub_event_msg->queue_simple_scrub_event_msg to a separate commit preceding your main change.  In general, *whenever* you can move a self-contained refactor to its own commit please do so especially if it touches a bunch of lines of code irrelevant to the rest of the PR.

Also, I think I probably suggested queue_simple_scrub_event_msg, but let's instead call it queue_scrub_event_msg_default_cost.

File: src/osd/OSD.h
Comment (athanatos) [inline src/osd/OSD.h:1628]: Same deal here, move the addition of this helper and supporting changes to a separate commit.

File: src/osd/scrubber/pg_scrubber.cc
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.cc:845]: The comment immediately above should indicate what this return value is.

Comment (athanatos) [reply src/osd/scrubber/pg_scrubber.cc:845]: Why int64_t?  You aren't using a negative value to encode an error (though that would be a valid choice).  It doesn't look like it's possible for this function to return a negative value.

File: src/osd/scrubber/pg_scrubber.cc
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.cc:945]: So, 0 if scrub is blocked, 1 if it's not blocked, but there are no further objects, and N otherwise?  That's pretty confusing -- how about std::optional<unsigned> where std::nullopt indicates blocked and otherwise it simply encodes the number of objects in the range?

Alternately, you could keep the return value a signed integer and use -EAGAIN to indicate that it's blocked.  @ronen-fr Do you have a preference?

File: src/osd/scrubber/pg_scrubber.cc
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.cc:966]: How about num_chunk_objects?  chunk_size could mean a few things.

File: src/osd/scrubber/pg_scrubber.cc
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.cc:976]: pg_avg_object_size

File: src/osd/scrubber/pg_scrubber.h
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.h:797]: Function comment should indicate *exactly* how the caller should interpret the return value.

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/osd/scrubber/pg_scrubber.cc
Comment (ronen-fr) [reply src/osd/scrubber/pg_scrubber.cc:845]: (that's an old and highly debated issue in "C++ circles". My preference is to follow the "do not use unsigned to anything else but bitmasks" - as Google says. But there are other views. A contrasting one is here:
https://youtu.be/82jVpEmAEV4?si=QJAjsd4QBXwcJYc5
)

File: src/osd/scrubber/pg_scrubber.cc
Comment (ronen-fr) [reply src/osd/scrubber/pg_scrubber.cc:945]: I'm definitely for the 'optional'

File: src/osd/scrubber/pg_scrubber.cc
Comment (athanatos) [reply src/osd/scrubber/pg_scrubber.cc:845]: I'd *really* rather types encode the actual range they are supposed to represent.

**Comment by amathuria** [comment]
jenkins test make check

**Comment by amathuria** [comment]
@athanatos @ronen-fr  does this look better to both of you?

**Comment by ronen-fr** [comment]
there are now two versions of get_scrub_cost() - one in the OSD class, and one in the PgScrubber.
Can you make sure, first, that both are still needed; then: rename one of them, and - add detailed function descriptions?

File: src/osd/scrubber/pg_scrubber.cc
Comment (ronen-fr) [inline src/osd/scrubber/pg_scrubber.cc:947]: maybe just:
return objects.size();

File: src/osd/scrubber/pg_scrubber.cc
Comment (ronen-fr) [inline src/osd/scrubber/pg_scrubber.cc:957]: !!!!! must not call select_range() again !!

File: src/osd/scrubber/pg_scrubber.cc
Comment (ronen-fr) [inline src/osd/scrubber/pg_scrubber.cc:968]: why pass 'is_deep', which is the value of a class member?

File: src/osd/scrubber/pg_scrubber.cc
Comment (ronen-fr) [inline src/osd/scrubber/pg_scrubber.cc:981]: pls use fmtlib for new code

File: src/osd/scrubber/pg_scrubber.h
Comment (ronen-fr) [inline src/osd/scrubber/pg_scrubber.h:732]: see other comment re passing is_deep

File: src/osd/OSD.cc
Comment (sseshasa) [inline src/osd/OSD.cc:1801]: `OSDService::get_scrub_cost()` is no longer used and can be removed as part of a standalone last commit.

**Comment by sseshasa** [comment]
@amathuria The commit sequence can be improved a bit as I found a couple of things confusing with respect to passing cost parameter to the various methods. Here's my suggestion on the sequence:

1. First commit - Use commit "osd: Add function to get osd_bandwidth_cost_per_io" as is.
2. Second commit - introduce `std::optional<uint64_t> PgScrubber::select_range()` and `uint64_t PgScrubber::get_scrub_cost(uint64_t num_chunk_objects, bool is_deep)` methods, but don't use them anywhere.
3. Third commit - Add the rest of the changes into this commit.
4. Fourth commit - Remove `OSDService::get_scrub_cost()`

I think the above would help maintain a good history for anyone looking into this PR at a later point. The above can of course be improved further as you see fit.

**Comment by amathuria** [comment]
> @amathuria The commit sequence can be improved a bit as I found a couple of things confusing with respect to passing cost parameter to the various methods. Here's my suggestion on the sequence:
> 
>     1. First commit - Use commit "osd: Add function to get osd_bandwidth_cost_per_io" as is.
> 
>     2. Second commit - introduce `std::optional<uint64_t> PgScrubber::select_range()` and `uint64_t PgScrubber::get_scrub_cost(uint64_t num_chunk_objects, bool is_deep)` methods, but don't use them anywhere.
> 
>     3. Third commit - Add the rest of the changes into this commit.
> 
>     4. Fourth commit - Remove `OSDService::get_scrub_cost()`
> 
> 
> I think the above would help maintain a good history for anyone looking into this PR at a later point. The above can of course be improved further as you see fit.

Thanks Sridhar, I'll change that.

**Comment by amathuria** [comment]
jenkins test api

File: src/osd/scrubber/pg_scrubber.cc
Comment (sseshasa) [inline src/osd/scrubber/pg_scrubber.cc:955]: Doesn't this latest change result in calling select_range() twice?

**Review by sseshasa — CHANGES_REQUESTED** [review]

File: src/osd/scrubber/pg_scrubber.cc
Comment (amathuria) [reply src/osd/scrubber/pg_scrubber.cc]: Sorry about that! Changed it now

Comment (sseshasa) [reply src/osd/scrubber/pg_scrubber.cc]: Considering that `num_chunk_objects` is not used outside the if..else block, you can consider making its scope tighter by using the if statement with initializer like I indicated originally. But in this case, since the method ends after the if..else block, I am okay if you leave it as is.

**Review by sseshasa — APPROVED** [review]
LGTM

**Comment by amathuria** [comment]
jenkins test make check

**Comment by amathuria** [comment]
@athanatos  could you take a look at this again?

File: src/osd/OSD.h
Comment (athanatos) [inline src/osd/OSD.h:121]: This function isn't always valid -- it crashes the OSD if called on wpq.  That's ok here, but please warn the next reader by adding a comment:
```
/// scheduler cost per io, only valid for mclock, asserts for wpq
double get_cost_per_io() const;
```

File: src/osd/OSD.cc
Comment (athanatos) [inline src/osd/OSD.cc]: Looks like this one wasn't fixed?

**Review by athanatos — APPROVED** [review]
LGTM once you've addressed my comments.

File: src/osd/OSD.cc
Comment (amathuria) [reply src/osd/OSD.cc]: It's fixed as part of the next commit, where we start using the new scrub cost.

**Comment by amathuria** [comment]
jenkins test api

**Comment by amathuria** [comment]
jenkins test windows

**Review by pereman2 — APPROVED** [review]
https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues65641 rados approved

**Pull request merged**
